blog/_posts/2018-03-29-missing.md

```diff
@@ -0,0 +1,394 @@
+`Array{Union{Missing,T}}` rather than `Array{Any}` objects, to provide the compiler
+with the necessary type information.
+
+The second one consists in using a **compact memory layout for arrays with `Union`s
+of bits types**. The standard `Array` type now uses an optimized memory layout for
+element types which are `Union` of bits types, i.e. immutable types which contain
+no references (see `isbits`). This includes `Missing` and basic types such as
+`Int`, `Float64`, `Complex{Float64}` and `Date`. When `T` is a bits type,
+`Array{Union{Missing,T}}` objects are internally represented as a pair of arrays
+of the same size: an `Array{T}` holding non-missing values and uninitialized memory
+for missing values; and an `Array{UInt8}` storing a *type tag* indicating whether
+each entry is `Missing` or `T`.
+
+This layout consumes slightly more memory than the sentinel approach, as the
+type tag part occupies one byte for each entry. But this overhead is reasonable:
+for example, the memory usage of an `Array{Union{Missing,Float64}}` is only
+12.5% higher than that of an `Array{Float64}`. Compared with the sentinel approach,
+it has the advantage of being fully generic (as detailed in the previous section).
+Actually, this mechanism can be used in other situations, for example with `Union{Nothing,Int}`
+(which is the element type of the array returned by `indexin` in Julia 0.7).
+It is also more correct, as the sentinel approach can fail in corner cases: for example,
+in R, `NA + NaN` returns `NA` but `NaN + NA` returns `NaN` due to floating point
+computation rules.
+
+Arrays of non-bits types with missing values are already represented as efficiently
+as their counterparts without missing values. Indeed, such arrays consist of
+pointers to the actual objects which live in a different memory area. Missing values
+can be represented as a special pointer just like non-missing values. This is notably
+the case for `Array{Union{Missing,String}}`.
+
+The efficient memory layout of `Array` in the presence of missing values makes it
+unnecessary to use dedicated array types like
+[`DataArray`](https://github.com/JuliaStats/DataArrays.jl). In fact, the layout
+of the `DataArray` type is very similar to that of `Array{Union{Missing,T}}`
+described above. The only difference is that it uses a `BitArray` rather than an
+`Array{UInt8}` to indicate whether a value is missing, therefore taking 1 bit per
+entry rather than 8 bits. Even if it consumes more memory, the `Array{UInt8}` mask
+approach is faster (at least in the current state of `BitArray`), and it generalizes
+to `Union`s of more than two types. However, we are aware that other implementations
+such as [PostgreSQL](https://www.postgresql.org/docs/9.5/static/storage-page-layout.html)
+or [Apache Arrow](https://arrow.apache.org/docs/memory_layout.html#null-bitmaps)
+use bitmaps equivalent to `BitArray`.
+
+## Safety and propagation by default
```

Comment:
Maybe call this "A data scientist's null" (maybe with subtitle like the current title), to aid skimability

Reply:
....and maybe point out at the very top that poor handling of missing values is a common source of bugs/errors in published work so it's critical to get right (with some links). I think that's a really important motivation for this whole project.

Reply:
I've added a link. Do you have pointers in fields other than economics?

Regarding the title, I'm not sure using "a data scientist's null" for this section is a good idea given that it applies to all sections (which is mentioned at the top): data scientists need generic, efficient and safe missing values.